src/algebra/jordan/triple.lean

```diff
@@ -0,0 +1,215 @@
+
+lemma rgsmul (a b : A) (z : ℤ) (c : A) : ⦃a, b, z•c⦄ = z•⦃a, b, c⦄ :=
+  by rw [comm, lgsmul, comm]
+
+lemma lneg (a b c : A) : ⦃-a, b, c⦄ = -⦃a, b, c⦄ :=
+by rw [←sub_eq_zero, sub_neg_eq_add, ←ladd, neg_add_self, lzero]
+
+lemma mneg (a b c : A): ⦃a, -b, c⦄ = -⦃a, b, c⦄ :=
+by rw [←sub_eq_zero, sub_neg_eq_add, ←madd, neg_add_self, mzero]
+
+lemma rneg (a b c : A): ⦃a, b, -c⦄ = -⦃a, b, c⦄ := by rw [comm, lneg, comm]
+
+lemma lsub (a b c d : A) : ⦃a - d, b, c⦄ = ⦃a, b, c⦄ - ⦃d, b, c⦄ :=
+by rw [eq_sub_iff_add_eq, ← ladd, sub_add_cancel]
+
+lemma msub (a b c d : A) : ⦃a, b - c, d⦄ = ⦃a, b, d⦄ - ⦃a, c, d⦄ :=
+by rw [eq_sub_iff_add_eq, ← madd, sub_add_cancel]
+
+
+lemma rsub (a b c d : A) : ⦃a, b, c - d⦄ = ⦃a, b, c⦄ - ⦃a, b, d⦄ :=
+by rw [comm, lsub, comm, comm d]
+
+lemma lr_bilinear (a₁ a₂ b c₁ c₂ : A) : ⦃a₁ + a₂, b, c₁ + c₂⦄ =
+  ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄ :=
+calc ⦃a₁ + a₂, b, c₁ + c₂⦄ = ⦃a₁, b, c₁ + c₂⦄ + ⦃a₂, b, c₁ + c₂⦄ : by rw ladd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁ + c₂⦄ : by rw radd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + (⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄) : by rw radd
+... = ⦃a₁, b, c₁⦄ + ⦃a₁, b, c₂⦄ + ⦃a₂, b, c₁⦄ + ⦃a₂, b, c₂⦄ : by rw ← add_assoc
+
+lemma polar (a b c : A) : ⦃a + c, b, a + c⦄ = ⦃a, b, a⦄ + 2•⦃a, b, c⦄ + ⦃c, b, c⦄ :=
+calc ⦃a + c, b, a + c⦄ = ⦃a, b, a⦄ + ⦃a, b, c⦄ + ⦃c, b, a⦄ + ⦃c, b, c⦄ :
+  by rw lr_bilinear a c b a c
+... = ⦃a, b, a⦄ + ⦃a, b, c⦄ + ⦃a, b, c⦄ + ⦃c, b, c⦄ : by rw comm c b a
+... = ⦃a, b, a⦄ + (⦃a, b, c⦄ + ⦃a, b, c⦄)  + ⦃c, b, c⦄ : by rw ← add_assoc
+... = ⦃a, b, a⦄ + 2•⦃a, b, c⦄ + ⦃c, b, c⦄ : by rw two_nsmul
+
+/-- Define the multiplication operator `D` -/
+@[simps] def D (a b : A) : add_monoid.End A :=
+{ to_fun := λ c, ⦃a, b, c⦄,
+  map_zero' := rzero _ _,
+  map_add' := radd _ _, }
+
+/-- Define the Quadratic operator `Q` -/
+@[simps] def Q (a c : A) : add_monoid.End A :=
+{ to_fun := λ b, ⦃a, b, c⦄,
+  map_zero' := mzero _ _,
+  map_add' := begin
+    intros,
+    rw madd _ _,
+    exact _inst_3,
+  end }
+
+/--
+For a in A, the map b → D a b is an additive monoid homomorphism from A to add_monoid.End A
+-/
+@[simps] def D_madd (a : A) : A  →+  add_monoid.End A :=
+{ to_fun := λ b, D a b,
+  map_zero' := begin
+    ext c,
+    rw [D_apply, mzero, add_monoid_hom.zero_apply],
+  end,
+  map_add' := λ b₁ b₂, begin
+    ext c,
+    rw [D_apply, madd, add_monoid_hom.add_apply, D_apply, D_apply],
+  end, }
+
+/--
+The map a → D a is an additive monoid homomorphism from A to (A  →+  add_monoid.End A)
+-/
+@[simps] def D_ladd : A  →+ (A  →+  add_monoid.End A) :=
+{ to_fun := λ a, D_madd a,
+  map_zero' := begin
+    ext b c,
+    rw [D_madd_apply, D_apply, add_monoid_hom.zero_apply,  add_monoid_hom.zero_apply, lzero ],
+  end,
+  map_add' := λ a₁ a₂, begin
+    ext b c,
+    rw [D_madd_apply, D_apply, add_monoid_hom.add_apply, D_madd_apply, D_madd_apply,
+      add_monoid_hom.add_apply, D_apply, D_apply, ladd],
+  end }
```

Comment:
I think it would be best to build this up all at once as `add_monoid_hom.tp : A →+ A →+ A →+ A`. You can then pull out the `D` and `Q` specializations using `flip` etc later.

Reply:
Something like:
```
@[simps] def add_monoid_hom.tp : A →+ A →+ A →+ A := 
{ to_fun := λ a, {
    to_fun := λ b, {
      to_fun := λ c, ⦃a, b, c⦄,
      map_zero' := sorry,
      map_add' := sorry, },
    map_zero' := sorry,
    map_add' := sorry, },
  map_zero' := sorry,
  map_add' := sorry, }
```
?

Reply:
Yes, exactly, The outermost sorries will be something like ` λ a₁ a₂, linear_map.ext $ λ b, linear_map.ext $ λ c, _` I think.

You could try to use `linear_map.mk\_2` for the inner two maps, but I don't think it will be any clearer

Reply:
Done, thanks.